omaha/handler.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package omaha
+
+import (
+	"encoding/xml"
+	"log"
+	"net/http"
+)
+
+type OmahaHandler struct {
+	Updater
+}
+
+func (o *OmahaHandler) ServeHTTP(w http.ResponseWriter, httpReq *http.Request) {
+	if httpReq.Method != "POST" {
+		log.Printf("omaha: Unexpected HTTP method: %s", httpReq.Method)
+		http.Error(w, "Expected a POST", http.StatusBadRequest)
+		return
+	}
+
+	// A request over 1M in size is certainly bogus.
+	reader := http.MaxBytesReader(w, httpReq.Body, 1024*1024)
+
+	decoder := xml.NewDecoder(reader)
+	var omahaReq Request
+	if err := decoder.Decode(&omahaReq); err != nil {
+		log.Printf("omaha: Failed decoding XML: %v", err)
+		http.Error(w, "Invalid XML", http.StatusBadRequest)
+		return
+	}
+
+	if omahaReq.Protocol != "3.0" {
+		log.Printf("omaha: Unexpected protocol: %q", omahaReq.Protocol)
+		http.Error(w, "Omaha 3.0 Required", http.StatusBadRequest)
+		return
+	}
+
+	httpStatus := 0
+	omahaResp := NewResponse()
+	for _, appReq := range omahaReq.Apps {
+		appResp := o.serveApp(omahaResp, httpReq, &omahaReq, appReq)
+		if appResp.Status == AppOK {
+			// HTTP is ok if any app is ok.
+			httpStatus = http.StatusOK
+		} else if httpStatus == 0 {
```

Comment:
if the 1st app is OK, and 2nd app is not-ok, the final result will still be "ok", is that desired?